components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/AddonsAdapter.kt

```diff
@@ -0,0 +1,218 @@
+        val addButton = view.findViewById<ImageView>(R.id.add_button)
+        return AddonViewHolder(
+            view,
+            iconView,
+            titleView,
+            summaryView,
+            ratingView,
+            userCountView,
+            addButton
+        )
+    }
+
+    override fun getItemCount() = items.size
+
+    override fun getItemViewType(position: Int): Int {
+        return when (items[position]) {
+            is Addon -> VIEW_HOLDER_TYPE_ADDON
+            is Section -> VIEW_HOLDER_TYPE_SECTION
+            is NotYetSupportedSection -> VIEW_HOLDER_TYPE_NOT_YET_SUPPORTED_SECTION
+            else -> throw IllegalArgumentException("items[position] has unrecognized type")
+        }
+    }
+
+    override fun onBindViewHolder(holder: CustomViewHolder, position: Int) {
+        val item = items[position]
+
+        when (holder) {
+            is SectionViewHolder -> bindSection(holder, item as Section)
+            is AddonViewHolder -> bindAddon(holder, item as Addon)
+            is UnsupportedSectionViewHolder -> bindNotYetSupportedSection(holder, item as NotYetSupportedSection)
+        }
+    }
+
+    private fun bindSection(holder: SectionViewHolder, section: Section) {
+        holder.titleView.setText(section.title)
+    }
+
+    private fun bindNotYetSupportedSection(holder: UnsupportedSectionViewHolder, section: NotYetSupportedSection) {
+        val context = holder.itemView.context
+        holder.titleView.setText(section.title)
+        holder.itemView.setOnClickListener {
+            onNotYetSupportedClicked(unsupportedAddons)
+        }
+    }
+
+    private fun bindAddon(holder: AddonViewHolder, addon: Addon) {
+        val context = holder.itemView.context
+        addon.rating?.let {
+            val userCount = context.getString(R.string.mozac_feature_addons_user_rating_count)
+            val ratingContentDescription =
+                context.getString(R.string.mozac_feature_addons_rating_content_description)
+            holder.ratingView.contentDescription =
+                String.format(ratingContentDescription, it.average)
+            holder.ratingView.rating = it.average
+            holder.userCountView.text = String.format(userCount, getFormattedAmount(it.reviews))
+        }
+
+        holder.titleView.text =
+            if (addon.translatableName.isNotEmpty()) {
+                addon.translatableName.translate()
+            } else {
+                addon.id
+            }
+
+        if (addon.translatableSummary.isNotEmpty()) {
+            holder.summaryView.text = addon.translatableSummary.translate()
+        } else {
+            holder.summaryView.visibility = View.GONE
+        }
+
+        holder.itemView.tag = addon
+        holder.itemView.setOnClickListener(clickListener)
+        holder.addButton.isVisible = !addon.isInstalled()
+        val listener = if (!addon.isInstalled()) {
+            clickListener
```

Comment:
I think we are setting the same listener when the an `add-on` item and the  `addButton` is clicked, we should have separated listeners.

When `add-on` item is clicked and  is not installed apps may want to show a details page when.
When `add-on` item is clicked and installed apps may want to show the add on settings page.
When `addButton` item is clicked apps want to start the installation process and show some UI to the user.

Reply:
You can see the different screens on this [video](https://drive.google.com/file/d/1nB5WGeyrqU7HpLDzWIwW7qR_4GoLDSwG/view?usp=sharing)

Reply:
Below is where we are handling the listener. As we are exposing this as reusable class I think we should be a more explicit. 

 ```kotlin
   override fun onClick(view: View) {
        val context = view.context
        when (view.id) {
            R.id.add_button -> {
                val addon = (((view.parent) as View).tag as Addon)
                showPermissionDialog(addon)
            }
            R.id.add_on_item -> {
                val addon = view.tag as Addon
                if (addon.isInstalled()) {
                    val intent = Intent(context, InstalledAddonDetailsActivity::class.java)
                    intent.putExtra("add_on", addon)
                    context.startActivity(intent)
                } else {
                    val intent = Intent(context, AddonDetailsActivity::class.java)
                    intent.putExtra("add_on", addon)
                    this.startActivity(intent)
                }
            }
            else -> {}
        }
    }
```

Reply:
As we have multiple callbacks we could create a delegate interface to reduce the number of parameters in the constructor, something like `AddonsManagerAdapterDelage`.

The interface could have functions like:
``` kotlin
fun onAddonItemClicked(addon:Addon) = Unit
fun onInstalledAddonButtonClicked(addon:Addon) = Unit
fun onNotYetSupportedSectionClicked(unSupportedAddons:(ArrayList<Addon>) ) = Unit
```

Reply:
Fixed.

Reply:
Awesome 🚀